### PR TITLE
Chrysalis Phase 2 completion

### PR DESCRIPTION
**Pull request by DeppLearning**
Successor of https://github.com/iotaledger/identity.rs/pull/155
Includes https://github.com/iotaledger/identity.rs/pull/160

### CODE REVIEW

File: examples/Cargo.toml [hunk 1]
```diff
@@ -7,8 +7,11 @@ publish = false
 [dependencies]
 identity = { path = "../identity" }
 rand = { version = "0.8" }
-smol = { version = "0.1", features = ["tokio02"] }
-smol-potat = { version = "0.3" }
+iota-core = { git = "https://github.com/iotaledger/iota.rs", rev = "86f5bb24727266e13abd772616d0700abe866892" }
+#iota-core = { git = "https://github.com/iotaledger/iota.rs", rev = "c993159e2a0d4eb250dc82e43e233219144e5931" }
```
Comment (l1h3r) [inline examples/Cargo.toml:11]: should remove all these unused deps

File: examples/create_did.rs [hunk 2]
```diff
@@ -3,14 +3,17 @@
...
 use identity::iota::Result;
 use identity::iota::TangleRef;
 
-#[smol_potat::main]
+#[tokio::main]
 async fn main() -> Result<()> {
-  let client: Client = Client::new()?;
+  // Create a new client connected to the Testnet (Chrysalis).
+  // Node-syncing has to be disabled for now.
+  let client: Client = ClientBuilder::new().node_sync_disabled().build().await?;
```
Comment (l1h3r) [inline examples/create_did.rs:16]: There should be a `Client::builder` function. very minor but can avoid an extra import

File: identity-iota/Cargo.toml [hunk 3]
```diff
@@ -12,12 +12,15 @@ homepage = "https://www.iota.org"
...
 [dependencies]
 async-trait = { version = "0.1", default-features = false }
+futures = { version = "0.3" }
 identity-core = { version = "=0.2.0", path = "../identity-core" }
 identity-credential = { version = "=0.2.0", path = "../identity-credential" }
 identity-did = { version = "=0.2.0", path = "../identity-did" }
 iota-constants = { version = "0.2", default-features = false }
 iota-conversion = { version = "0.5", default-features = false }
-iota-core = { git = "https://github.com/Thoralf-M/iota.rs", rev = "d7c8c64fc3ac2340f0148708a916c245f42fd454" }
+#iota-core = { git = "https://github.com/Thoralf-M/iota.rs", rev = "d7c8c64fc3ac2340f0148708a916c245f42fd454" }
```
Comment (l1h3r) [inline identity-iota/Cargo.toml:21]: unused deps

File: identity-iota/src/tangle/message_index.rs [hunk 4]
```diff
@@ -143,55 +142,55 @@ mod tests {
   }
 
   #[rustfmt::skip]
```
Comment (l1h3r) [inline identity-iota/src/tangle/message_index.rs:144]: these `#[rustfmt::skip]` attributes should have been removed a long time ago. probably best to do now

File: identity-iota/src/lib.rs [hunk 5]
```diff
@@ -31,7 +31,7 @@ pub mod did;
 pub mod error;
 pub mod tangle;
 
-pub(crate) mod utils;
+// pub(crate) mod utils;
```
Comment (l1h3r) [inline identity-iota/src/lib.rs:34]: unused?

File: identity-iota/src/client/client_builder.rs [hunk 6]
```diff
@@ -41,9 +43,14 @@ impl ClientBuilder {
     self
   }
 
+  pub fn node_sync_disabled(mut self) -> Self {
```
Comment (l1h3r) [inline identity-iota/src/client/client_builder.rs:46]: `pub fn node_sync_enabled(mut self, value: bool) -> Self` leaves much more flexibility for the caller

File: identity-iota/src/client/client_builder.rs [hunk 7]
```diff
@@ -1,23 +1,25 @@
...
 
 impl ClientBuilder {
   /// Creates a new `ClientBuilder`.
-  pub const fn new() -> Self {
+  pub fn new() -> Self {
     Self {
-      network: Network::Mainnet,
+      network: Network::default(),
       nodes: Vec::new(),
+      node_sync_enabled: true,
```
Comment (l1h3r) [inline identity-iota/src/client/client_builder.rs:22]: can this default to `false` since it seems to be required at the moment?

probably also good to use a constant

File: identity-iota/src/client/client_builder.rs [hunk 8]
```diff
@@ -1,23 +1,25 @@
...
-use crate::client::Client;
-use crate::client::Network;
-use crate::error::Result;
+use crate::{
+  client::{Client, Network},
+  error::Result,
+};
 
 /// A `ClientBuilder` is used to generated a customized `Client`.
-#[derive(Clone, Debug)]
```
Comment (l1h3r) [inline identity-iota/src/client/client_builder.rs:9]: why this change? `bool` should be debuggable

File: identity-iota/src/client/client.rs [hunk 9]
```diff
@@ -1,44 +1,38 @@
...
-use crate::error::Result;
-use crate::tangle::Message;
-use crate::tangle::MessageId;
-use crate::utils::bundles_from_trytes;
-use crate::utils::create_address_from_trits;
-use crate::utils::encode_trits;
-use crate::utils::txn_hash_trytes;
+use futures::stream::FuturesUnordered;
+use futures::stream::StreamExt;
+use identity_core::{common::Url, convert::ToJson};
```
Comment (l1h3r) [inline identity-iota/src/client/client.rs:6]: should consistently use the same style of import.

some form of consistency will be enforced through `rustfmt.toml` eventually

**Comment by l1h3r** [comment]
Looks like CI is failing because the wasm bindings don't compile after these changes - just need to handle the change to `async` here:

https://github.com/iotaledger/identity.rs/blob/dev/bindings/wasm/src/iota.rs#L45

File: identity-iota/src/client/client.rs (hunk 9 above)
Comment (l1h3r) [reply identity-iota/src/client/client.rs:6]: ahh I see this was the style from my old PR 😄 should move it to the new style

File: identity-iota/src/client/client_builder.rs [hunk 10]
```diff
@@ -1,23 +1,25 @@
...
 pub struct ClientBuilder {
   pub(crate) network: Network,
   pub(crate) nodes: Vec<String>,
+  pub(crate) node_sync_enabled: bool,
 }
 
 impl ClientBuilder {
   /// Creates a new `ClientBuilder`.
-  pub const fn new() -> Self {
```
Comment (l1h3r) [inline identity-iota/src/client/client_builder.rs:17]: any reason `const` was dropped?

File: identity-iota/src/client/network.rs [hunk 11]
```diff
@@ -2,15 +2,14 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use identity_core::common::Url;
-use iota::client::builder;
 
 use crate::did::DID;
 
 lazy_static! {
-  static ref EXPLORER_MAIN: Url = Url::parse("https://explorer.iota.org/mainnet").unwrap();
+  static ref EXPLORER_MAIN: Url = Url::parse("https://explorer.iota.org/chrysalis").unwrap();
```
Comment (l1h3r) [inline identity-iota/src/client/network.rs:9]: We are dropping the `Comnet` variant I thought?

I don't really see a reason to keep the old networks. Definitely think it's a bit confusing to mix them both.

I think just `enum Network { Main, Test }` will do?

cc @JelleMillenaar

File: identity-iota/src/client/client_builder.rs (hunk 10 above)
Comment (DeppLearning) [reply identity-iota/src/client/client_builder.rs:17]: Removed this to be able to call Network::default. Reverted to Network::main now and added that const again.

File: identity-iota/src/client/client_builder.rs (hunk 8 above)
Comment (DeppLearning) [reply identity-iota/src/client/client_builder.rs:9]: Opps, this is a left over change from when we briefly went with something like:
```
pub struct ClientBuilder {
  pub(crate) builder: iota::ClientBuilder,
  pub(crate) network: Network,
}
``` 
before noticing that it's probably better not to wrap it in the struct. Reverted this

**Comment by l1h3r** [comment]
Currently blocked by lack of wasm support in iota.rs

https://github.com/iotaledger/iota.rs/issues/444

**Review by l1h3r — APPROVED** [review]
